From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vansencool <vansencool@gmail.com>
Date: Thu, 25 Sep 2025 13:27:45 +0530
Subject: [PATCH] Lazy regeneration


diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
index 8578d1f78ddd1bb75f3230f04bfaa35af9f5f822..4a0264b28f6f95df0ce717c60e71ee81fd899e1e 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -7,6 +7,7 @@ import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.lighting.LevelLightEngine;
 
@@ -40,6 +41,13 @@ public class ClientboundLevelChunkWithLightPacket implements Packet<ClientGamePa
         this.x = pos.x;
         this.z = pos.z;
         io.papermc.paper.antixray.ChunkPacketInfo<net.minecraft.world.level.block.state.BlockState> chunkPacketInfo = modifyBlocks ? chunk.getLevel().chunkPacketBlockController.getChunkPacketInfo(this, chunk) : null; // Paper - Ant-Xray
+        // NeonPaper start - Lazy regeneration
+        if (!Level.regeneratedChunks.contains(chunk.getPos().longKey) && Level.pendingRegeneration.containsKey(chunk.getPos().longKey)) {
+            Level.setChunkAt(chunk, Level.pendingRegeneration.get(chunk.getPos().longKey));
+            Level.pendingRegeneration.remove(chunk.getPos().longKey);
+            Level.regeneratedChunks.remove(chunk.getPos().longKey);
+        }
+        // NeonPaper end - Lazy regeneration
         this.chunkData = new ClientboundLevelChunkPacketData(chunk, chunkPacketInfo); // Paper - Anti-Xray
         this.lightData = new ClientboundLightUpdatePacketData(pos, lightEngine, skyLight, blockLight);
         chunk.getLevel().chunkPacketBlockController.modifyBlocks(this, chunkPacketInfo); // Paper - Anti-Xray - Modify blocks
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 45a79eb10ccdd8559d5ce3e0cc7c3b070d584d32..c7f9b26db6537bde66da26acb7c6b2433d9d6f67 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -168,6 +168,7 @@ import net.minecraft.world.level.storage.WorldData;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 import net.vansen.neonpaper.auto.AutoRegeneration;
+import net.vansen.neonpaper.regeneration.PendingChunks;
 import org.slf4j.Logger;
 
 public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements ServerInfo, ChunkIOErrorReporter, CommandSource, ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer { // Paper - rewrite chunk system
@@ -1015,6 +1016,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
         }
 
+        // NeonPaper start - Lazy regeneration
+        PendingChunks.shutdown();
+        // NeonPaper end - Lazy regeneration
+
         LOGGER.info("Saving worlds");
 
         for (ServerLevel serverLevel : this.getAllLevels()) {
diff --git a/net/minecraft/server/level/ChunkHolder.java b/net/minecraft/server/level/ChunkHolder.java
index 656041c9539b6834b4d37b353eb6b810a7763ff4..9649bbf21148a5b5e7dab7f0f81ba79ccdb7c92e 100644
--- a/net/minecraft/server/level/ChunkHolder.java
+++ b/net/minecraft/server/level/ChunkHolder.java
@@ -106,6 +106,13 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
     public final LevelChunk moonrise$getFullChunk() {
         if (this.newChunkHolder.isFullChunkReady()) {
             if (this.newChunkHolder.getCurrentChunk() instanceof LevelChunk levelChunk) {
+                // NeonPaper start - Lazy regeneration
+                if (Level.pendingRegeneration.containsKey(levelChunk.getPos().longKey)) {
+                    Level.setChunkAt(levelChunk, Level.pendingRegeneration.get(levelChunk.getPos().longKey));
+                    Level.pendingRegeneration.remove(levelChunk.getPos().longKey);
+                    Level.regeneratedChunks.add(levelChunk.getPos().longKey);
+                }
+                // NeonPaper end - Lazy regeneration
                 return levelChunk;
             } // else: race condition: chunk unload
         }
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 6540b2d6a1062d883811ce240c49d30d1925b291..f8410a2acacae27869cf14d0cccaf2a04279bb59 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -85,6 +85,13 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     @Override
     public final void moonrise$setFullChunk(final int chunkX, final int chunkZ, final LevelChunk chunk) {
         final long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        // NeonPaper start - Lazy regeneration
+        if (chunk != null) {
+            if (Level.pendingRegeneration.containsKey(chunk.getPos().longKey)) {
+                Level.regenerateChunk(chunk);
+            }
+        }
+        // NeonPaper end - Lazy regeneration
         if (chunk == null) {
             this.fullChunks.remove(key);
         } else {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 034153d5517fa57649f9213896ddeabd431f4878..c1974405967e69581e8d7cdefafae2976d62cafb 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -4,11 +4,16 @@ import com.google.common.collect.Lists;
 import com.mojang.serialization.Codec;
 import java.io.IOException;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.FeatureHooks;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -26,8 +31,11 @@ import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import io.papermc.paper.util.MCUtil;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ChunkHolder;
 import net.minecraft.server.level.FullChunkStatus;
+import net.minecraft.server.level.ServerChunkCache;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
@@ -158,6 +166,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public Map<BlockPos, BlockEntity> capturedTileEntities = new java.util.LinkedHashMap<>(); // Paper - Retain block place order when capturing blockstates
     public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
     public final it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<SpawnCategory> ticksPerSpawnCategory = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>();
+    // NeonPaper start - Lazy regeneration
+    public static ConcurrentHashMap<Long, SnappedChunk> pendingRegeneration = new ConcurrentHashMap<>();
+    public static Set<Long> regeneratedChunks = new HashSet<>();
+    // NeonPaper end - Lazy regeneration
     // Paper start
     public int wakeupInactiveRemainingAnimals;
     public int wakeupInactiveRemainingFlying;
@@ -1100,7 +1112,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         setChunkAt(getChunk(x, z), chunk);
     }
 
-    public static PalettedContainer<BlockState> packed(PalettedContainer<BlockState> src) {
+    private static PalettedContainer<BlockState> packed(PalettedContainer<BlockState> src) {
         return PalettedContainer.unpack(
                 src.registry,
                 src.strategy,
@@ -1112,7 +1124,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         });
     }
 
-    public PalettedContainer<BlockState> direct(PalettedContainer<BlockState> src) {
+    private static PalettedContainer<BlockState> direct(PalettedContainer<BlockState> src) {
         Palette<BlockState> palette = src.data.palette();
         int size = palette.getSize();
         List<BlockState> paletteValues = new java.util.ArrayList<>(size);
@@ -1131,7 +1143,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         );
     }
 
-    public void setChunkAt(LevelChunk currentChunk, SnappedChunk chunk) {
+    public static void setChunkAt(ChunkAccess currentChunk, SnappedChunk chunk) {
         int count = currentChunk.getSectionsCount();
         for (int i = 0; i < count; ++i) {
             LevelChunkSection section = currentChunk.getSection(i);
@@ -1151,6 +1163,19 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
         currentChunk.markUnsaved();
     }
+
+    public static void regenerateChunk(LevelChunk chunk) {
+        Level.setChunkAt(chunk, Level.pendingRegeneration.get(chunk.getPos().longKey));
+        Level.pendingRegeneration.remove(chunk.getPos().longKey);
+        Level.regeneratedChunks.add(chunk.getPos().longKey);
+        ChunkHolder playerChunk = ((ServerChunkCache) chunk.getLevel().getChunkSource()).chunkMap.getVisibleChunkIfPresent(chunk.getPos().longKey);
+        if (playerChunk == null) return;
+
+        List<ServerPlayer> playersInRange = playerChunk.playerProvider.getPlayers(playerChunk.getPos(), false);
+        if (playersInRange.isEmpty()) return;
+
+        FeatureHooks.sendChunkRefreshPackets(playersInRange, chunk);
+    }
     // NeonPaper end
 
     @Override
